### PR TITLE
Optimize `ecovec![]` macro

### PR DESCRIPTION
**Pull request by Kmeakin**
LLVM is not smart enough to eliminate calls to `EcoVec::reserve()` even when they are unecessary thanks to earlier calls to `EcoVec::reserve()` or `EcoVec::with_capacity()`. This patch elimates them when the `ecovec![]` macro is used. There may be other unnecessary calls which can be elimated in future patches.

`ecovec![1, 2, 3]` results in 3 unecessary calls to `EcoVec::reserve()`. Fixed by forwarding to the `From<[T; N]>` impl.

`ecovec![1; 42]` results in 42 unecessary calls to `EcoVec::reserve()`. Fixed by adding `EcoVec::push_unchecked()` function for when we know resizing is unnecessary.

### CODE REVIEW

File: src/vec.rs [hunk 1]
```diff
@@ -241,12 +236,27 @@ impl<T: Clone> EcoVec<T> {
...
+    }
+
+    /// Add a value at the end of the vector, without reallocating.
+    ///
+    /// # Safety
+    /// Caller must ensure that `self.len < self.capacity()` and
+    /// `self.is_unique()` hold, such as by calling `EcoVec::reserve()`, or
+    /// `EcoVec::with_capacity()`.
+    #[inline]
+    pub unsafe fn push_unchecked(&mut self, value: T) {
```
Comment (laurmaedje) [inline src/vec.rs:250]: I wonder whether this method should really be public. `Vec` does not have this method. Can we make it private?

**Comment by laurmaedje** [comment]
Originally, I planned to make a release today so that `EcoVec::new` constness is available for use through crates.io. If you plan to submit further PRs, I would wait with this.

File: src/vec.rs (hunk 1 above)
Comment (Kmeakin) [reply src/vec.rs:250]: I don't see any harm in it. It is marked `unsafe`, so any misuse is the caller's responsibility.

Comment (laurmaedje) [reply src/vec.rs:250]: Personally, I think a method should only be public if there's a strong case for it being useful. The fact that `Vec` does not expose such a method indicates to me that it is not typically needed. I think it's a good goal to keep a crate's public API surface as minimal as possible.

**Pull request merged**